### PR TITLE
Update Godoc for ReadyDoneAware

### PR DESCRIPTION
**Pull request by smnzhu**
This PR updates the Godoc for `ReadyDoneAware` to more clearly define the conventions for its implementation. There are many different implementations in the code base which interpret the `Ready` and `Done` methods to mean different things (see https://github.com/dapperlabs/flow-go/issues/5702). 

In particular, the implementations are inconsistent regarding the concurrency-safety and idempotency of these methods. It is also not clear by looking at the interface whether the `Ready` / `Done` methods actually provide the mechanism to start / stop the module, or simply return channels that are closed when the module is started / stopped via some other means (in the latter case, how does one start / stop the module?).

Documenting the interface will hopefully lessen the confusion, and encourage better implementations in the future.

### CODE REVIEW

File: module/common.go [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
-// ReadyDoneAware provides easy interface to wait for module startup and shutdown
+// ReadyDoneAware provides an easy interface to wait for module startup and shutdown.
+// Modules that implement this interface only support a single start-stop cycle, and
+// will not restart if Ready() is called again after shutdown has already commenced.
 type ReadyDoneAware interface {
+	// Ready commences startup of the module, and returns a ready channel that is closed once
+	// startup has completed.
+	// If shutdown has already commenced before this method is called for the first time,
+	// startup will not be performed and the returned channel will never close.
+	// This is an idempotent method.
```
Comment (huitseeker) [inline module/common.go:11]:
```suggestion
	// This should be an idempotent method.
```

File: module/common.go [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
 	Ready() <-chan struct{}
+
+	// Done commences shutdown of the module, and returns a done channel that is closed once
+	// shutdown has completed.
+	// This is an idempotent method.
```
Comment (huitseeker) [inline module/common.go:16]:
```suggestion
	// This should be an idempotent method.
```

**Review by huitseeker — APPROVED** [review]
Thanks! a couple of nits!

**Review by vishalchangrani — COMMENTED** [review]
❤️ thanks for doing this!

**Review by vishalchangrani — APPROVED** [review]

**Pull request merged**